bundles/org.openhab.core.model.sitemap/src/org/openhab/core/model/sitemap/Sitemap.xtext

```diff
@@ -137,6 +137,7 @@ Colorpicker:
 
 Input:
     'Input' (('item=' item=ItemRef) & ('label=' label=(ID | STRING))? & ('icon=' icon=Icon)? &
+    ('inputHint=' inputHint=(STRING))? &
```

Comment:
Maybe limit the possible values in the Sitemap syntax? Otherwise syntax issues aren't visible in the server log.

Reply:
This is possible, but also not done for any of the other parameters. I would like to see some broader opinion on this before making an exception for this parameter.

Reply:
Good remark. For what other existing parameters should have been expected?

Reply:
In general parser only checks if any parameter is an int, ID, string, float or boolean. The parser itself does not check if e.g. an item is valid, although that would show up in other places when trying to update the non existing item from a sitemap UI.
Likewise, there is no check if icons are valid, colors are valid.
The chart period parameter has a limited set of allowed values, no check in the parser.
The video encoding parameter as a limited list of options.
There is no validation for the url parameters.

All of these will fail at runtime, Not when saving the sitemap.

Reply:
Is the parenthesis around STRING required ?
There are sometimes there, sometimes not.
For example the url attribute has no parenthesis in the Image element but the url attribute has parenthesis in the Video element.
The two syntax are identical or is there a real difference ?

Reply:
> Is the parenthesis around STRING required ?
> There are sometimes there, sometimes not.
> For example the url attribute has no parenthesis in the Image element but the url attribute has parenthesis in the Video element.
> The two syntax are identical or is there a real difference ?

I don’t think the parentheses are required, but didn’t test.

Reply:
Please also note that there are some controls added outside the .xtext file. They are defined in package "validation".

Reply:
> For example the url attribute has no parenthesis in the Image element but the url attribute has parenthesis in the Video element.

I removed the extra parentheses from in a few places in the file.

When doing that, I noticed that the url parameter for the Image and Video parameter are different. It is a required parameter for the Video, but not for the Image. Is that correct? There is no difference in the documentation.

Reply:
The image can be set as:
* base64 encoded item state
* URL

The video must be provided by URL.

Reply:
Here is what is written in the documentation:

> `url` is the default URL from which to retrieve the image, if there is no associated Item or if the associated item's state is not a URL.

In practice, BasicUI is properly handling the case url is not set or not set to a valid URL.
@mueller-ma : can you confirm that it is also handled in the Android App ?

So yes, `url` is optional for image while mandatory for video.

Maybe the documentation could be slightly updated.

Reply:
Thinking more about it, I really think the syntax should consider/check the allowed values when it is clear that only few values are accepted. It could be done either directly in the xtext file or in "validation" class.
It could help users to identify some mistakes in their sitemaps (because they used an unexpected value).

I am not asking to fix all the existing syntax in that PR but I would like we do it in a next PR.
Maybe we can at least start with the `inputHint`. Are you ok @mherwege ?

Reply:
> can you confirm that it is also handled in the Android App ?

Yes, see [here](https://github.com/openhab/openhab-android/blob/main/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt#L676-682).

Reply:
The Android app checks:
1. is widget state a base64 encoded image => decode and use it
2. is widget.url set => use it
3. else => don't display an image

https://github.com/openhab/openhab-android/blob/main/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt#L676-L683

Reply:
Thank @mueller-ma  and @maniac103 , it confirms `url` is really optional for image.

Reply:
Yes, I believe it is very easy to do it directly in the grammar. Look at BOOLEAN_OBJECT rule as an example.

Reply:
I added a validation in the grammar. I got this in the log when using an illegal value in a sitemap, and the sitemap was not accepted:
```
Configuration model 'thuis.sitemap' has errors, therefore ignoring it: [13,67]: mismatched input '"test"' expecting RULE_INPUT_HINT
```

Reply:
Looks good.